File: engine/mmra_engine/petrel_grv.py
"""
petrel_grv.py – GRV from Petrel structural + fluid-contact marginal cases (3 + 3).

When Petrel provides three GRV values for depth/structure (at fixed mid contact) and
three for fluid contact (at fixed mid structure), build a 3×3 GRV matrix and sample
depth × contact cases each Monte Carlo iteration.

Combination (default): multiplicative from P50 anchor
    GRV[d,c] = GRV_mid × (depth_grv[d] / depth_grv[mid]) × (contact_grv[c] / contact_grv[mid])
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional, Sequence, Tuple

import numpy as np

CASE_LABELS = ("P90", "P50", "P10")


@dataclass
class PetrelGrvMarginals:
    """Three structural GRVs (mid contact) + three contact GRVs (mid structure)."""

    depth_grv: List[float]
    contact_grv: List[float]
    depth_weights: List[float] = field(default_factory=lambda: [1 / 3, 1 / 3, 1 / 3])
    contact_weights: List[float] = field(default_factory=lambda: [1 / 3, 1 / 3, 1 / 3])
    grv_matrix_3x3: Optional[List[List[float]]] = None

    def __post_init__(self) -> None:
        if len(self.depth_grv) != 3 or len(self.contact_grv) != 3:
            raise ValueError("depth_grv and contact_grv must each have 3 values (P90, P50, P10).")
        if len(self.depth_weights) != 3 or len(self.contact_weights) != 3:
            raise ValueError("depth_weights and contact_weights must each have 3 values.")
        _validate_positive(self.depth_grv, "depth_grv")
        _validate_positive(self.contact_grv, "contact_grv")
        _normalize_weights(self.depth_weights)
        _normalize_weights(self.contact_weights)


def _validate_positive(values: Sequence[float], name: str) -> None:
    for i, v in enumerate(values):
        if v is None or not np.isfinite(v) or v <= 0:
            raise ValueError(f"{name}[{i}] must be a positive finite number.")


def _normalize_weights(weights: List[float]) -> None:
    s = float(sum(weights))
    if s <= 0:
        raise ValueError("Case weights must sum to a positive value.")
    for i in range(len(weights)):
        weights[i] = float(weights[i]) / s


def petrel_marginals_from_dict(data: Optional[dict]) -> Optional[PetrelGrvMarginals]:
    if not data:
        return None
    return PetrelGrvMarginals(
        depth_grv=[float(x) for x in data["depth_grv"]],
        contact_grv=[float(x) for x in data["contact_grv"]],
        depth_weights=[float(x) for x in data.get("depth_weights", [1 / 3, 1 / 3, 1 / 3])],
        contact_weights=[float(x) for x in data.get("contact_weights", [1 / 3, 1 / 3, 1 / 3])],
        grv_matrix_3x3=data.get("grv_matrix_3x3"),
    )


def petrel_marginals_to_dict(pm: Optional[PetrelGrvMarginals]) -> Optional[dict]:
    if pm is None:
        return None
    out = {
        "depth_grv": list(pm.depth_grv),
        "contact_grv": list(pm.contact_grv),
        "depth_weights": list(pm.depth_weights),
        "contact_weights": list(pm.contact_weights),
    }
    if pm.grv_matrix_3x3 is not None:
        out["grv_matrix_3x3"] = pm.grv_matrix_3x3
    return out


def build_grv_matrix_3x3(pm: PetrelGrvMarginals) -> List[List[float]]:
    """Build or return 3×3 GRV (acre-ft internal; caller converts units before/after)."""
    if pm.grv_matrix_3x3 is not None:
        if len(pm.grv_matrix_3x3) != 3 or any(len(r) != 3 for r in pm.grv_matrix_3x3):
            raise ValueError("grv_matrix_3x3 must be 3×3.")
        return [[float(c) for c in row] for row in pm.grv_matrix_3x3]

    d_mid = pm.depth_grv[1]
    c_mid = pm.contact_grv[1]
    if d_mid <= 0 or c_mid <= 0:
        raise ValueError("P50 depth_grv and contact_grv must be positive.")

    matrix: List[List[float]] = []
    for d_grv in pm.depth_grv:
        row: List[float] = []
        for c_grv in pm.contact_grv:
            row.append(float(d_grv * c_grv / c_mid))
        matrix.append(row)
    return matrix


def depth_marginal_grv(matrix: List[List[float]], case_index: int) -> float:
    """Mean GRV across contact cases for one depth row (tornado / reporting)."""
    return float(np.mean(matrix[case_index]))


def contact_marginal_grv(matrix: List[List[float]], case_index: int) -> float:
    """Mean GRV across depth cases for one contact column."""
    return float(np.mean([matrix[d][case_index] for d in range(3)]))


def sample_petrel_grv(
    pm: PetrelGrvMarginals,
    n: int,
    rng: np.random.Generator,
) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """
    Sample GRV from Petrel marginals.

    Returns
    -------
    grv_samples, depth_case_idx, contact_case_idx, grv_matrix (flat reference)
    """
    matrix = build_grv_matrix_3x3(pm)
    d_idx = rng.choice(3, size=n, p=np.asarray(pm.depth_weights, dtype=float))
    c_idx = rng.choice(3, size=n, p=np.asarray(pm.contact_weights, dtype=float))
    grv = np.array([matrix[int(d)][int(c)] for d, c in zip(d_idx, c_idx)], dtype=float)
    return grv, d_idx.astype(int), c_idx.astype(int), np.asarray(matrix, dtype=float)


def petrel_base_grv_scalar(pm: PetrelGrvMarginals) -> float:
    """P50×P50 cell for OAT base case."""
    matrix = build_grv_matrix_3x3(pm)
    return float(matrix[1][1])


def petrel_depth_swing_scalars(pm: PetrelGrvMarginals) -> Tuple[float, float, float]:
    """Low / base / high GRV for depth driver (contact averaged per row)."""
    matrix = build_grv_matrix_3x3(pm)
    return (
        depth_marginal_grv(matrix, 0),
        float(matrix[1][1]),
        depth_marginal_grv(matrix, 2),
    )


def petrel_contact_swing_scalars(pm: PetrelGrvMarginals) -> Tuple[float, float, float]:
    """Low / base / high GRV for fluid-contact driver."""
    matrix = build_grv_matrix_3x3(pm)
    return (
        contact_marginal_grv(matrix, 0),
        float(matrix[1][1]),
        contact_marginal_grv(matrix, 2),
    )

File: frontend/src/utils/petrelGrv.ts
import type { PetrelGrvMarginals, RockVolumeInputUnit, SimulationInput } from '../types/api'

export const PETREL_CASE_LABELS = ['P90 (low)', 'P50 (mid)', 'P10 (high)'] as const

/**
 * Three percentile case column headers for the Petrel 3+3 GRV matrix.
 *
 * These are the low/mid/high cases of EACH GRV axis (depth and fluid contact),
 * not the Setup GRV-uncertainty source names. Stored vectors are ordered
 * [low, mid, high] = [P90, P50, P10].
 */
export function petrelCaseLabels(): [string, string, string] {
  return ['P90', 'P50', 'P10']
}

/** Row label for the structural (depth) GRV axis in Petrel 3+3 (from Setup GRV uncertainties). */
export function petrelDepthRowLabel(grvParamLabels: string[]): string {
  return grvParamLabels[0]?.trim() || 'Structure (depth)'
}

/** Row label for the fluid-contact GRV axis in Petrel 3+3 (from Setup GRV uncertainties). */
export function petrelContactRowLabel(grvParamLabels: string[]): string {
  return grvParamLabels[2]?.trim() || grvParamLabels[1]?.trim() || 'Fluid contact'
}

export function defaultPetrelGrvMarginals(
  depthMid = 10_000,
  contactMid = 10_000,
): PetrelGrvMarginals {
  return {
    depth_grv: [depthMid * 0.8, depthMid, depthMid * 1.2],
    contact_grv: [contactMid * 0.9, contactMid, contactMid * 1.1],
    depth_weights: [1 / 3, 1 / 3, 1 / 3],
    contact_weights: [1 / 3, 1 / 3, 1 / 3],
  }
}

/** Build 3×3 GRV matrix from Petrel marginals (multiplicative combination). */
export function buildPetrelGrvMatrix(pm: PetrelGrvMarginals): number[][] {
  const [d90, d50, d10] = pm.depth_grv
  const [c90, c50, c10] = pm.contact_grv
  if (d50 <= 0 || c50 <= 0) throw new Error('P50 GRV values must be positive.')
  return [
    [d90 * c90 / c50, d90 * c50 / c50, d90 * c10 / c50],
    [d50 * c90 / c50, d50, d50 * c10 / c50],
    [d10 * c90 / c50, d10 * c50 / c50, d10 * c10 / c50],
  ]
}

export function ensurePetrelGrvMarginals(input: SimulationInput): SimulationInput {
  if (input.petrel_grv_marginals) return input
  return {
    ...input,
    nrv_entry_mode: 'petrel_marginals',
    petrel_grv_marginals: defaultPetrelGrvMarginals(),
  }
}

export function patchPetrelMarginals(
  input: SimulationInput,
  patch: Partial<PetrelGrvMarginals>,
): SimulationInput {
  const current = input.petrel_grv_marginals ?? defaultPetrelGrvMarginals()
  return {
    ...input,
    petrel_grv_marginals: { ...current, ...patch },
  }
}

export function updatePetrelCaseValue(
  pm: PetrelGrvMarginals,
  axis: 'depth' | 'contact',
  caseIndex: number,
  value: number,
): PetrelGrvMarginals {
  const key = axis === 'depth' ? 'depth_grv' : 'contact_grv'
  const next = [...pm[key]] as [number, number, number]
  next[caseIndex] = value
  return { ...pm, [key]: next }
}

export function petrelMatrixSummary(pm: PetrelGrvMarginals, unit: RockVolumeInputUnit): string {
  try {
    const m = buildPetrelGrvMatrix(pm)
    return `Derived 3×3 GRV (${unit}), mid cell = ${m[1][1].toLocaleString()}`
  } catch {
    return 'Check Petrel GRV values.'
  }
}
